Add per-lookup per-syllable flag

behdad opened the pull request:
Fixes https://github.com/harfbuzz/harfbuzz/issues/3513

jfkthame reviewed (approved):
LGTM!

jfkthame commented:
Would be good to add the Lohit example, at least, as a test.

behdad commented:
> Would be good to add the Lohit example, at least, as a test.

Yes I'll add. Thanks.

dscorbett reviewed (changes requested):
This PR does not match Uniscribe.

I made a font with:
```fea
feature blws {
    sub uni0B95 uni0B95 by uni0B950B95;
    sub u11315 u11315 by u11315_u11315;
} blws;
```
I made another font with the same substitutions in 'ss01'. I tested them in Internet Explorer. With both fonts, the Tamil sequence failed to ligate, and the Grantha sequence ligated. I infer that, in Uniscribe, there is no distinction between script-specific presentation features and general presentation features. The Tamil shaper restricts both to syllables; USE restricts neither to syllables. That is not what this PR does.

behdad commented:
So basically just remove the clear_syllables?

behdad commented:
Okay if we remove the clear_syllables in Indic, then all features, including GPOS, will be constrained. I'm fine with that.

khaledhosny commented:
Was not that a known Uniscribe limitation that HarfBuzz and CoreText lifted, or am I confusing this with something else?

behdad commented:
> Was not that a known Uniscribe limitation that HarfBuzz and CoreText lifted, or am I confusing this with something else?

I think you are right...

dscorbett commented:
I added a lookup:
```fea
feature dist {
    pos uni0B95 uni0B95 -300;
} dist;
```
Uniscribe applied that lookup for Tamil. Syllables should therefore be cleared before GPOS.

NorbertLindenberg commented:
> Okay if we remove the clear_syllables in Indic, then all features, including GPOS, will be constrained. I'm fine with that.

What about the rclt feature?
http://tiro.com/John/Fixing_Indic2_OTL.pdf

dscorbett commented:
I changed 'blws' to 'rclt' and got the same results. It looks like Uniscribe’s Tamil shaper does not support multi-cluster substitutions.

behdad commented:
Do we know what CoreText does?

NorbertLindenberg commented:
Does rclt work as John's document proposes in CoreText and older HarfBuzz? I thought these two had implemented it, while Microsoft and Adobe didn't.

dscorbett commented:
Core Text behaves like HarfBuzz 4.1.0. It applies presentation features (including 'rclt') across clusters, and it shapes “दिसं” wrong in Lohit Devanagari.

behdad commented:
Okay, which behavior do we want? @dscorbett & @jfkthame I defer to you to decide.

behdad commented:
I have implementation of all three different possibilities ready.

behdad commented:
Alternate fix, matching Uniscribe, is in https://github.com/harfbuzz/harfbuzz/pull/3516
One of our previous tests breaks with that change. Would be worth checking that the new output matches Uniscribe at least.

behdad commented:
@tiroj would you be okay with changing HarfBuzz to constrain user features to within syllable in the Indic shaper?

dscorbett reviewed (commented):
Since Uniscribe and Core Text are inconsistent with each other, I guess there is not much additional harm in HarfBuzz being inconsistent with them both. If you don’t mind that, then I think this PR is a fine approach.

behdad commented on src/hb-ot-shape-complex-khmer.cc:
```diff
@@ -110,13 +110,7 @@ collect_features_khmer (hb_ot_shape_planner_t *plan)
   map->enable_feature (HB_TAG('l','o','c','l'));
   map->enable_feature (HB_TAG('c','c','m','p'));
```
I retained how it was before, which IIRC was intentional; ie. I remember in testing we had the `setup_syllables` before `locl` and `ccmp` in USE because I belive Uniscribe had it that way.

behdad replied:
Oh. I see what you mean. I should add to others as well.

behdad replied:
Fixed.

jfkthame commented:
> Okay, which behavior do we want? @dscorbett & @jfkthame I defer to you to decide.

My vote (though I don't feel really strongly about it) would be to constrain the Indic features (as explicitly listed in the spec) but allow arbitrary user features to work across clusters.

Given what the spec says, and the established implementation in Uniscribe, I think we *have* to constrain the Indic features because fonts may quite reasonably be relying on this (like Lohit Devanagari). But for arbitrary features that aren't part of the Indic specs, removing the restriction allows the font designer greater flexibility, and this seems unlikely to cause significant breakage. (Older fonts that were targeted primarily at the Uniscribe engine aren't likely to make much use of additional features, as most Uniscribe-based software didn't provide any mechanism for the user to apply them AFAIK.)

behdad commented:
> > Okay, which behavior do we want? @dscorbett & @jfkthame I defer to you to decide.

Okay so both of you are in favor of this PR. I'm happy with that. And I've added the Lohit testcase now, and addressed David's review. I think it's ready to be merged.

tiroj commented:
Minimum should be application of rclt across cluster boundaries (and all GPOS, of course), as agreed in April 2014. Anything beyond that _could_ break existing fonts if they were made to conform with cluster limitation behaviour, whether consciously or not.

Not allowing any cross-cluster GSUB is terrible, and I am still giving Microsoft a hard time about it. Some of the Indic scripts need things like head stroke extensions contextually inserted to create space for marks between adjacent above stroke shapes. Without being able to use cross-cluster context, that just doesn’t work.

So long as I have one GSUB feature that is a) processed in the general typographic block by lookup order and b) works reliably across clusters, I am happy.

behdad commented:
I think what we have now confirms to those requirements. Testing would be appreciated though.  We'll get a release out soon.
